embedded-graphics/MIGRATING-0.5-0.6.md

```diff
@@ -0,0 +1,422 @@
+    pixelcolor::{
+        raw::{RawData, RawU1},
+        BinaryColor, PixelColor,
+    },
+    DrawTarget,
+};
+
+impl<C> DrawTarget<C> for DisplayDriver
+where
+    C: PixelColor + Into<BinaryColor>,
+{
+    type Error = DI::Error;
+
+    fn draw_pixel(&mut self, pixel: drawable::Pixel<C>) -> Result<(), Self::Error> {
+        let drawable::Pixel(pos, color) = pixel;
+
+        // Guard against negative values. All positive i32 values from `pos` can be represented in
+        // the `u32`s that `set_pixel()` accepts...
+        if pos.x < 0 || pos.y < 0 {
+            return Ok(());
+        }
+
+        // ... which makes the `as` coercions here safe.
+        self.set_pixel(
+            pos.x as u32,
+            pos.y as u32,
+            // Convery color to BinaryColor, then into underlying raw storage u8
+            color.into().into_storage(),
+        );
+
+        Ok(())
+    }
+
+    fn size(&self) -> Size {
+        let (w, h) = self.get_dimensions();
+
+        Size::new(w as u32, h as u32)
+    }
+}
+```
+
+Some notes on the above:
+
+- The `Drawing` trait is renamed to `DrawTarget`.
+- This implementation takes a `C: PixelColor + Into<BinaryColor>` allowing items using other color types to be used on the monochrome display. If this is not possible or undesired, use the appropriate pixel type directly. For example:
+
+  ```rust
+  impl DrawTarget<Rgb565> for DisplayDriver {
+      // ...
+  }
+  ```
+
+- The `draw()` method is replaced by `draw_pixel()`. This method should handle setting of individual pixels on the display. How it does this is at the descretion of the display driver (pixel buffer, immediate mode, etc).
+- A new `size()` method is now a required item. This should return the width and height of the display as a `Size`.
+- The implementation of `DrawTarget` must now provide an associated `Error` type. In the above example `core::convert::Infallible` is used, however a better error type should be used to communicate hardware failures, etc to the user.
+- Any pixels that are offscreen (negative coordinates or greater than display dimensions) should result in a noop. In the example above, `self.set_pixel()` (defined by the SSD1306 driver) will not attempt to set any pixels beyond the positive screen limits.
+- `draw_pixel()` now returns `Result<(), Self::Error>` to account for driver error handling.
+
+### Choosing the right pixel color type
+
+Below are some common use cases to help choose the right pixel color for a given display module. The full list of pixel colors is described in [the table above](#pixel-colors).
+
+#### `BinaryColor`
+
+If the display only supports two states, use `BinaryColor`. This is applicable to monochrome OLED displays like the SSD1306 or SH1106, character/bitmap LCDs and even LED matrices.
+
+#### `Gray2`
+
+Use for displays that can represent 2 or 3 color states. For example, tricolor epaper displays that can show white, black or red like the SSD1675 should use this type.
```

Comment:
I don't think this is correct. `Gray2` is a 2 bpp color, so there are 4 color states and this shouldn't be used for 2 color states. Tricolor epaper display could also use an eum, as shown in the example: https://github.com/jamwaffles/embedded-graphics/blob/272346369e312068cc504c079f0b4ce7ffa32a8a/embedded-graphics/src/pixelcolor/mod.rs#L7-L75

Maybe `Gray4` is a better example, because it could be used for `SSD132x` displays.

Reply:
I ended up removing this example. I couldn't think of a better one to replace it with.